### PR TITLE
Allow AMA AttorneyTasks to be cancelled

### PR DESCRIPTION
**Pull request by hschallhorn**
Connects #12115

### Description
Adds task action to attorney task to allow them to be cancelled and sent back to the assigning judge to be assigned again.

Screen Shot 2019-10-07 at 11 46 57 AM
Screen Shot 2019-10-07 at 11 47 04 AM
Screen Shot 2019-10-07 at 11 47 15 AM
Screen Shot 2019-10-07 at 11 47 28 AM

### Acceptance Criteria
- [ ] Attorneys can cancel AMA `AttorneyTask`s assigned to them

### Testing Plan
1. Sign in as "BVASCASPER1"
2. Go to any assigned task in the user's queue
3. Ensure they have the 'Cancel Task' action
4. Ensure cancelling the task removes it from the user's queue and sends the task back to the judge

### CODE REVIEW

File: spec/models/tasks/attorney_task_spec.rb [hunk 1]
```diff
@@ -83,13 +85,29 @@
...
-        Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h
+        Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
+        Constants.TASK_ACTIONS.CANCEL_TASK.to_h
       ]
 
       expect(subject).to eq(expected_actions)
     end
   end
+
+  context "when cancelling the task" do
```
Comment (lomky) [inline spec/models/tasks/attorney_task_spec.rb]: :100:!

File: app/models/tasks/attorney_task.rb [hunk 2]
```diff
@@ -55,4 +58,21 @@ def assigned_to_role_is_valid
   def assigned_by_role_is_valid
     errors.add(:assigned_by, "has to be a judge") if assigned_by && !assigned_by.judge_in_vacols?
   end
+
+  def task_just_cancelled?
```
[the remark this answers is not in the document]
Comment (hschallhorn) [reply app/models/tasks/attorney_task.rb]: Great point. It does not look like there is any place in the code where we cancel attorney tasks.

File: app/models/tasks/attorney_task.rb [hunk 3]
```diff
@@ -14,16 +14,22 @@ class AttorneyTask < Task
   validate :assigned_to_role_is_valid
   validate :child_attorney_tasks_are_completed, on: :create
 
+  after_update :send_back_to_judge_assign, if: :task_just_cancelled?
+
   def available_actions(user)
     if parent.is_a?(JudgeTask) && parent.assigned_to == user
```
Comment (lomky) [inline app/models/tasks/attorney_task.rb]: The way we do the Judge tasks, this lets the Judge who is the Decision Review judge (`parent.assigned_to`) cancel the task. But in the case that the Assigning Judge was different, they don't have the ability to cancel. Do we want to try to explicitly enable the Assigning Judge's ability to cancel?

Comment (hschallhorn) [reply app/models/tasks/attorney_task.rb]: Great point! As the code works now, the [assigning judge is also the reviewing judge](https://github.com/department-of-veterans-affairs/caseflow/blob/5fc7d88144b0003e17daed97d1cacfd5977973e0/app/controllers/judge_assign_tasks_controller.rb#L33). Judge tasks can be [reassigned](https://github.com/department-of-veterans-affairs/caseflow/blob/e82803c2cdb863b53ece9796ab1a2585739d1e5b/app/models/tasks/judge_task.rb#L13), but this happens before an attorney task is created. Also, an attorney can return a decision to a different judge, but this "reassignment" occurs after the attorney task is completed.

Comment (lomky) [reply app/models/tasks/attorney_task.rb]: Looks like we can reassign JudgeDecisionReviewTasks that are `on_hold` (i.e. assigned out to an attorney). We should be able to determine the assigning judge via the `assigned_by` on the AttorneyTask

**Pull request merged**